attestation-agent/attester/src/cca/mod.rs

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2023 Arm Ltd.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use super::Attester;
+use anyhow::*;
+use base64::Engine;
+use nix::fcntl::{open, OFlag};
+use nix::sys::stat::Mode;
+use nix::unistd::close;
+use serde::{Deserialize, Serialize};
+use std::path::Path;
+
+const CCA_DEVICE_PATH: &str = "/dev/cca_attestation";
+
+// NOTE: The path might be different when the CCA feature is public available, will come back to update the actual path if needed.
+pub fn detect_platform() -> bool {
+    Path::new(CCA_DEVICE_PATH).exists()
+}
+
+#[derive(Debug, Default)]
+pub struct CCAAttester {}
+
+#[derive(Serialize, Deserialize)]
+struct CcaEvidence {
+    /// CCA token
+    token: Vec<u8>,
+}
+
+#[allow(non_camel_case_types)]
+#[repr(C)]
+pub struct cca_ioctl_request {
+    challenge: [u8; 64],
+    token: [u8; 4096],
+    token_length: u64,
+}
+
+nix::ioctl_readwrite!(cca_attestation_request, b'A', 1, cca_ioctl_request);
+
+#[async_trait::async_trait]
+impl Attester for CCAAttester {
+    async fn get_evidence(&self, mut challenge: Vec<u8>) -> Result<String> {
+        challenge.resize(64, 0);
+        let token = attestation(&challenge)?;
+        let evidence = CcaEvidence { token };
+        let ev = serde_json::to_string(&evidence).context("Serialize CCA evidence failed")?;
+        Ok(ev)
+    }
+}
+
+fn attestation(challenge: &[u8]) -> Result<Vec<u8>, Error> {
+    log::info!("cca_test::attestation started");
+
+    match open(CCA_DEVICE_PATH, OFlag::empty(), Mode::empty()) {
+        Result::Ok(f) => {
+            log::info!("cca_test::attestation opening attestation succeeded");
+            let mut request = cca_ioctl_request {
+                challenge: [0u8; 64],
+                token: [0u8; 4096],
+                token_length: 0u64,
+            };
+
+            let mut i: usize = 0;
+            let j: usize = std::cmp::min(request.challenge.len(), challenge.len());
+            while i < j {
+                request.challenge[i] = challenge[i];
+                i += 1;
+            }
```

Comment:
we can directly set `challenge` field with 
```rust
challenge.resize(64, 0);
... 

let request = cca_ioctl_request {
challenge,
...
}
challenge,
```

also change the api `fn attestation(&[u8])` to `fn attestation(mut Vec<u8>)`

in this way we can avoid copying byte by byte manually

Reply:
@Xynnn007 is this what you meant?

```patch
--- a/attestation-agent/attester/src/cca/mod.rs
+++ b/attestation-agent/attester/src/cca/mod.rs
@@ -42,32 +42,29 @@ nix::ioctl_readwrite!(cca_attestation_request, b'A', 1, cca_ioctl_request);
 impl Attester for CCAAttester {
     async fn get_evidence(&self, mut challenge: Vec<u8>) -> Result<String> {
         challenge.resize(64, 0);
-        let token = attestation(&challenge)?;
+        let token = attestation(challenge)?;
         let evidence = CcaEvidence { token };
         let ev = serde_json::to_string(&evidence).context("Serialize CCA evidence failed")?;
         Ok(ev)
     }
 }

-fn attestation(challenge: &[u8]) -> Result<Vec<u8>, Error> {
+fn attestation(challenge: Vec<u8>) -> Result<Vec<u8>, Error> {
     log::info!("cca_test::attestation started");

+    let challenge: [u8; 64] = challenge.try_into().unwrap_or_else(|v: Vec<u8>| {
+        panic!("Expected challenge to be length 64 but it was {}", v.len())
+    });
+
     match open(CCA_DEVICE_PATH, OFlag::empty(), Mode::empty()) {
         Result::Ok(f) => {
             log::info!("cca_test::attestation opening attestation succeeded");
             let mut request = cca_ioctl_request {
-                challenge: [0u8; 64],
+                challenge,
                 token: [0u8; 4096],
                 token_length: 0u64,
             };

-            let mut i: usize = 0;
-            let j: usize = std::cmp::min(request.challenge.len(), challenge.len());
-            while i < j {
-                request.challenge[i] = challenge[i];
-                i += 1;
-            }
-

Reply:
```
+    challenge.resize(64, 0);
+    let challenge = challenge.as_slice().try_into()?;
```

Reply:
thanks @Xynnn007 !

besides, I saw there is workflow failed (Run cargo test), but it can passed locally on my desktop, is that matter or anything related with change?

Reply:
here it is: 
https://github.com/confidential-containers/guest-components/actions/runs/6065048631/job/16545468411#step:11:471